File: Cpp/A101-memoizedFibonacci.cpp
#include <vector>
#include <iostream>
#include <chrono>

using std::vector,std::cout,std::chrono::high_resolution_clock,std::chrono::duration;

template<typename Func, typename... Args>
auto measureTime(Func&& func, Args&&... args){
    auto start = high_resolution_clock::now();
    auto result = std::forward<Func>(func)(std::forward<Args>(args)...);
    auto end = high_resolution_clock::now();
    duration<double,std::milli> duration = end - start;

    cout << "Execution time: " << duration.count() << " ms." << std::endl;

    return result;
};


class MemoizedFibonacci{
private:
    vector<long long> fibonacci;
    int steps;

    long long calculateFibonacciIter(int n){
        while(fibonacci.size()<=n){
            long long newFib = fibonacci[fibonacci.size()-2] + fibonacci[fibonacci.size()-1];
            fibonacci.push_back(newFib);
        };
        return fibonacci[n];
    };

public:
    MemoizedFibonacci(const vector<long long>& preCalc = {0,1}): fibonacci(preCalc),steps(0){
        cout << "Initial fibonacci Sequence:";
        for (const auto& num: fibonacci){
            cout << num << " ";
        }
        cout << std::endl;
    }
    // auto constructor
    long long calculate(int n){
        return measureTime([this,n](){return calculateFibonacciIter(n);});
    }
    vector<long long> getFibonacci(){
        return fibonacci;
    }
};

int main(){
    const int k=10;
    MemoizedFibonacci memFib = MemoizedFibonacci();
    memFib.calculate(k);
    for(auto num:memFib.getFibonacci()){
        cout << num << ",";
    };
    cout << std::endl;
}

File: TypeScript/A101-memoizedFibonacci.ts
function measureTime<T>(func: (...args:any[])=>T): (...args:any[])=>T{
    return function(...args:any[]):T{
        const start = performance.now();
        const result= func(...args);
        const total = performance.now() - start;
        console.log(`Execution time: ${total.toFixed(3)}ms.`)
        return result
    }
}

class memoizedFibonacci{
    private fibonacci:number[]=[0,1];
    private steps: number = 0;
    constructor(preCalc?:number[]){
        if (preCalc && preCalc.length>this.fibonacci.length){
            this.fibonacci=preCalc;
        };
        console.log(this.fibonacci);
    }
    private calculateFibonacciIter(n:number):number{
        while(this.fibonacci.length<=n){
            const newFib = this.fibonacci[this.fibonacci.length-2]+this.fibonacci[this.fibonacci.length-1]
            this.fibonacci.push(newFib)
        }
        return this.fibonacci[n]
    }
    calculate(n:number):number{
        const measuredFib=measureTime<number>(this.calculateFibonacciIter.bind(this))
        return measuredFib(n)
    }
    get_fibonacci():number[]{
        return this.fibonacci
    }
}

const n=50
const memFib = new memoizedFibonacci()
memFib.calculate(n)
console.log(memFib.get_fibonacci())
